### PR TITLE
Observable interface for subject second attempt

### PR DESCRIPTION
**Pull request by escamoteur**
Would be great if you could merge soon :-)

### CODE REVIEW

**Comment by escamoteur** [comment]
I had to change one test because it expected to find an Observable and not a Stream. But I think that should not be a Problem. I expect most people in future to directly subscribe to the Subject and not accessing the .stream property

**Comment by escamoteur** [comment]
https://github.com/ReactiveX/rxdart/issues/141

File: lib/src/subjects/subject.dart [hunk 1]
```diff
@@ -11,15 +11,12 @@ import 'package:rxdart/src/observable.dart';
 /// Please see `PublishSubject` for the simplest example of how to
 /// extend from this class, or `BehaviorSubject` for a slightly more
 /// complex example.
-abstract class Subject<T> implements StreamController<T> {
+abstract class Subject<T> extends Observable<T> implements StreamController<T> {
```
Comment (frankpepermans) [inline lib/src/subjects/subject.dart]: I'd change this to:
```
Subject(this.controller, Observable<T> observable) : super(observable);

  @override
  Observable<T> get stream => super.stream as Observable<T>;
```

File: lib/src/subjects/subject.dart [hunk 2]
```diff
@@ -11,15 +11,12 @@ import 'package:rxdart/src/observable.dart';
...
   final StreamController<T> controller;
   final Observable<T> observable;
   bool _isAddingStreamItems = false;
 
-  Subject(this.controller, this.observable);
-
-  @override
-  Observable<T> get stream => observable;
```
Comment (frankpepermans) [inline lib/src/subjects/subject.dart]: Keep the getter, but use:
```
@override
  Observable<T> get stream => super.stream as Observable<T>;
```

File: lib/src/subjects/subject.dart [hunk 3]
```diff
@@ -11,15 +11,12 @@ import 'package:rxdart/src/observable.dart';
 /// Please see `PublishSubject` for the simplest example of how to
 /// extend from this class, or `BehaviorSubject` for a slightly more
 /// complex example.
-abstract class Subject<T> implements StreamController<T> {
+abstract class Subject<T> extends Observable<T> implements StreamController<T> {
   final StreamController<T> controller;
   final Observable<T> observable;
```
Comment (frankpepermans) [inline lib/src/subjects/subject.dart]: remove this line, as Observable already exposes ```stream```

File: test/transformers/do_test.dart [hunk 4]
```diff
@@ -30,7 +30,7 @@ void main() {
         () async {
       int count = 0;
       final BehaviorSubject<int> subject = new BehaviorSubject<int>(sync: true);
-      final Stream<int> stream = subject.stream.doOnError(
```
Comment (frankpepermans) [inline test/transformers/do_test.dart]: You can leave the stream getter if you apply the above comments

**Comment by escamoteur** [comment]
Good idea

File: lib/src/subjects/subject.dart (hunk 3 above)
Comment (escamoteur) [reply lib/src/subjects/subject.dart]: which line do you mean?

File: lib/src/subjects/subject.dart (hunk 2 above)
Comment (escamoteur) [reply lib/src/subjects/subject.dart]: Did this but now there is the warning we should avoid `as`

Comment (frankpepermans) [reply lib/src/subjects/subject.dart]: Try making the stream property covariant in Observable

Comment (escamoteur) [reply lib/src/subjects/subject.dart]: I fear I don't know how you mean that? I didn't want to touch the Observable class if possible

Comment (escamoteur) [reply lib/src/subjects/subject.dart]: OK read now about what covariance means will try but if you could give me hint would be great

Comment (escamoteur) [reply lib/src/subjects/subject.dart]: I now tried to find a solution for one hour without success. I give up.
The solution with `as` works. if you have a better idea please help me.

Comment (frankpepermans) [reply lib/src/subjects/subject.dart]: Had a quick play with ```covariant```, but couldn't solve it cleanly either, we could just let ```Observable``` be as-is, and update ```Subject``` as following:
```
Subject(this.controller, Observable<T> observable)
      : this._observable = observable,
        super(observable);

  @override
  Observable<T> get stream => _observable;
```
Not sure if this is the cleanest solution though, I'll check again later on.

Comment (escamoteur) [reply lib/src/subjects/subject.dart]: ok, will change it like that

**Comment by brianegan** [comment]
LGTM

**Pull request merged**